File: Algo/Leetcode/206ReverseLinkedList.cpp
//1. Keep the head on the first one. Move the afterwards one to the front one by one. O(n).
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head==NULL) return head;
        ListNode* res = head;
        while(head->next!=NULL){
            ListNode* temp = head->next;
            head->next = head->next->next;
            temp -> next = res;  //Make to the first place.
            res = temp;
        }
        return res;
    }
};

// pre cur two pointer. easy to understand
class Solution {
public:
    ListNode* reverseList(ListNode* cur) {
        ListNode* pre = NULL;
        ListNode* tmp;

        while(cur){
            tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }
};

File: Algo/Leetcode/206ReverseLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tail = head
        if head is None:
            return head

        while tail.next:
            tmp = tail.next
            tail.next = tail.next.next
            tmp.next = head
            head = tmp

        return head
